contracts/EtherCrowd.sol

```diff
@@ -114,7 +119,56 @@ contract EtherCrowd {
 
 
 
+    // ChainLink UpKeep part, maybe putting it in another file?
+
+    function checkUpkeep(bytes calldata /*checkData*/) external override view returns (bool upkeepNeeded, bytes memory /*performData*/) {
+        // This function will be executed Off chain by the Keeper node, this is why the computation is done here in order to save on gas fees
+        upkeepNeeded = (block.timestamp - lastCheck) > checkInterval;
+        // We don't use the checkData in this example. The checkData is defined when the Upkeep was registered.
+    }
+
+    function performUpkeep(bytes calldata  /*performData*/ ) override external {
+        lastCheck = block.timestamp;
+        checkCrowdsales();
+
+        // We don't use the performData in this example. The performData is generated by the Keeper's call to your checkUpkeep function
+    }   
+
+    function checkCrowdsales() private {
+        for(uint i=0; i < currentId; i++){
+            // looping through all the crowdsales
+            Crowdsale memory crowdsale = idToCrowdsale[i];
+            if(block.timestamp >= crowdsale.endDate){
+                // It means the crowdsale has ended
+                endCrowdsale(crowdsale);
+            }
+        }
+    }
+
+    function endCrowdsale(Crowdsale memory _crowdsale) private {
+        // TODO: Emit events
+        // Crowdsale has ended, refund the money or give it to the owner according to the situation
+        if(_crowdsale.currentAmount >= _crowdsale.goalAmount){
+            // The crowdsale is successful, we can transfer the money to the owner
+            payable(_crowdsale.owner).transfer(_crowdsale.currentAmount);
+            
+        }
+        else{
+            // refunding the investors
+            refund(_crowdsale);
+        }
+
+    }
+
+    function refund(Crowdsale memory _crowdsale) private {
+            for (uint i=0; i< _crowdsale.contributors.length ; i++) {
+                address contributor = _crowdsale.contributors[i];
```

Comment:
Rename to contributorAddress to improve readability.

Reply:
Will do!